### ISSUE TITLE
[NEW PASS] Consider adding Optimization Driven Quantum Circuit Reduction pass

### ISSUE THREAD

**Issue by bachase** [unitaryfoundation/ucc#368]
Consider adding Optimization Driven Quantum Circuit Reduction from [arxiv:2502.14715](https://arxiv.org/abs/2502.14715)

> Implementing a quantum circuit on specific hardware with a reduced available gate set is often associated with a substantial increase in the length of the equivalent circuit. This process is also known as transpilation and due to decoherence, it is mandatory to keep quantum circuits as short as possible, without affecting functionality. In this work we propose three different transpilation approaches, based on a localized term-replacement scheme, to substantially reduce circuit lengths while preserving the unitary operation implemented by the circuit. The first variant is based on a stochastic search scheme, and the other variants are driven by a database retrieval scheme and a machine learning based decision support. We show that our proposed methods generate short quantum circuits for restricted gate sets, superior to the typical results obtained by using different qiskit optimization levels. Our method can be applied to different gate sets and scales well with an arbitrary number of qubits.

The authors might have code to share as the paper completes the review process.

**Comment by bachase** [comment]
Awesome to hear @jbolns  -- definitely share anything in this issue to chat async, and we can also make it a topic at an upcoming community call.

**Comment by jbolns** [comment]
Mine is at a different segment of quantum. ZX, lattice surgery, compiling. Very Fowler-like type of thinking. Related to his research. And collaborating with him on it.
[https://github.com/jbolns/topologiq](https://github.com/jbolns/topologiq)

The link I see with the challenge in this GH issue is mainly on the use of graphs and algorithms together, with lessons-learned on significant challenges maintaining the "meaning" of things and scaling up things as size of graph increases.

But, big differences to consider: 
- Different algorithms. I have been working on the "spatial" side of things. They're going for a more statistical approach. 
- The underlying scientific basis is different (the sections I left out are the more quantum quantum sections).

Sum. I don't think I could tackle it entirely by myself but perhaps I could help with code implementation.

**Comment by jordandsullivan** [comment]
@jbolns Thanks for the outlining the context and your capacity on this one. What do you think of reaching out the authors of the paper and seeing if they were interested in a collaboration?

**Comment by ACE07-Sev** [comment]
Greetings there,

Hope all are well. I read both papers ([[1](https://arxiv.org/pdf/[2502.14715](https://arxiv.org/pdf/2502.14715))] and [[2](https://arxiv.org/pdf/[2307.07353](https://arxiv.org/pdf/2307.07353))]) and thought it would be best to share the premise of the paper's algorithm and make sure you actually want it added (I got most of it down, so I won't really need the author code).

The work is split into two categories:
- Singleton gates: These are constant gates, i.e., X, Y, Z, T, CX, etc.
- Parameterized gates: These are parameterized gates, i.e., RX, RY, RZ, etc.

Unlike qiskit's transpiler which contracts gates together, forming a unitary, and resynthesizes it in the target gate set, this approach needs to find the rotation angles itself using QML, which adds considerable extra cost. I'll elaborate on this in a moment.

**Overall Idea**
The work utilizes compute graphs (a graph where a node represents a circuit, and the edges leading from the root to that node form the gates that need to be applied in that order to produce said node) to find shorter circuits to replace a random slice of the circuit to be optimized and by doing so shorten its overall depth. All methods discussed rely on this compute graph/database being there, and simply propose different ways of querying from it.

**Why would compute graphs be problematic?**
Compute graphs are essentially an elegant way to represent different circuits we can make using the available operators/gates given to us by randomly generating them. This generation is done within a maximum graph depth to avoid getting memkilled. Even with this limitation in place, the resulting circuits produced are in the hundreds of thousands for 3 qubits, and these blow up significantly as we increase the number of qubits.

So, one has to make one of two choices:
- **Store the unitary as the nodes are generated:** This approach saves time in re-contraction of the gates but introduces significant memory overhead due to having to store hundreds of thousands to millions of numpy arrays of `np.complex128`.
- **Lazy evaluation of the unitaries:** This approach saves memory but requires re-contraction each time so it introduces considerably longer runtime.

Considering the massive cost this database generation incurs (and I would honestly only justify it as a pre-generated thing for constant qubit size like `qml.datasets` in pennylane, NOT as something to be generated for each circuit each time) the generation and manual query to check if a desired unitary is available makes it infeasible on larger than ~10 qubit case.

**Singletons vs Parameterized gates**
This is only for singletons. For the general case including parameterized gates this requires a QML fine-tuning per node to check if a node can come close enough to the desired unitary. That means potentially millions of fine-tuning calls, spanning hours if not days per call to this compilation pass. Assuming the same target gateset, and same number of qubits, this still will need to be re-trained, re-contracted, and re-checked for each circuit even with such constraints in place.

**Final Thoughts**
I would argue that the paper makes itself redundant as if we are to use QML to fine-tune an ansatz to fit a desired unitary, we might as well just do that from the get-go with a brickwall ansatz. Spanning massive set of circuits, manually training each one, and then checking each one for whether it's close enough all for a slight reduction in circuit depth feels a bit unjustified. Overall it feels like rolling multiple dices and hoping to find a shorter circuit in a short depth without many nodes being generated, and hoping for a smooth, easy to train unitary for the parameterized case.

Spending time on what the evolved version of this is, aka rewrite rules + qml fine-tuning, would be a better use of your resources and likely a better avenue if we are to challenge qiskit's transpiler performance.

Link below shows what the transpiler actually does, so it's easier to compare:
https://quantum.cloud.ibm.com/docs/en/api/qiskit/transpiler

That being said, if you want this knowing the result it will likely have, I can get working on it.

**Comment by bachase** [comment]
Thanks for that summary @ACE07-Sev! I think if someone is still interested, we are open to seeing that implementation and contribution. But as you note, it might not be worth the overhead. Extensions to use ML to effectively compress that circuit database would be interesting too, as there might be a reduced representation even if approximate. So in short, this is part of what UCC is all about -- evaluating and understanding these new research approaches, even if they don't work out.

**Comment by jbolns** [comment]
@jordandsullivan . Apologies for the late reply, I was on holidays.

I like the rewards system and may see it as a motivation to engage with other shorter-term issues. For reasons below, I see this issue as a "slow" matter to engage out of intellectual curiosity and remain interested irrespective of the existence of a potential monetary reward.

Having said that, it is clear @ACE07-Sev would likely be able to resolve this much quicker than I would want to.

@bachase @ACE07-Sev . My view of this issue has evolved since my last message.

Brad and I covered a few of the notes in Steve's reply in a conversation some weeks ago. The encoding challenge is not particularly worrisome. The graph needs to be manipulated variously, but the information in the specific nodes is stable. Ergo, the conclusion you arrive at is robust: the challenge is doing it efficiently. 

But that's what makes the issue interesting.

Once the foundational algorithm is in place and correctness is ensured, there is a need to start adding one or another kind of heuristics to make it more efficient. The process of finding these heuristics is probably more important than the foundational algorithm. For context, the first implementation of the ZX graph algorithm I am working on separately took 30 sec for a Steane circuit, and it's now down to 0.3 sec.

For the same reason, I maintain my desire for author involvement (will write again later this week to see if they're back from holidays) and wouldn't want to hurry a solution for the specific issue.
